### ISSUE TITLE
"security_socket_xxxx" events sometimes have wrong "port" values (connect/accept)

### ISSUE THREAD

**Issue by rafaeldtinoco** [aquasecurity/tracee#3730]
## Description

![image](https://github.com/aquasecurity/tracee/assets/7395852/9706c24a-96f1-4b14-8d4b-a0f10a5dadda)

**Comment by rafaeldtinoco** [comment]
Well only enabling the events for me I start getting events with `sin_port == 0`.

```
{"timestamp":1702952866110010631,"threadStartTime":1702947555895062735,"processorId":7,"processId":752615,"cgroupId":4576,"threadId":752639,"parentProcessId":1,"hostProcessId":752615,"hostThreadId":752639,"hostParentProcessId":1,"userId":1000,"mountNamespace":4026531841,"pidNamespace":4026531836,"processName":"ThreadPoolForeg","executable":{"path":""},"hostName":"rugged","containerId":"","container":{},"kubernetes":{},"eventId":"735","eventName":"security_socket_connect","matchedPolicies":[""],"argsNum":3,"returnValue":0,"syscall":"connect","stackAddresses":null,"contextFlags":{"containerStarted":false,"isCompat":false},"threadEntityId":1484817851,"processEntityId":2039491600,"parentEntityId":1975426032,"args":[{"name":"sockfd","type":"int","value":2},{"name":"type","type":"string","value":"SOCK_DGRAM"},{"name":"remote_addr","type":"struct sockaddr*","value":{"sa_family":"AF_INET","sin_addr":"104.244.42.2","sin_port":"0"}}]}
```

The `security_socket_connect()` probe also runs before:

```C
err = READ_ONCE(sock->ops)->connect(sock, (struct sockaddr *)address, addrlen, sock->file->f_flags | file_flags);
```

the socket protocol connect handler (tcp_connect here). It could a value coming from userspace but I don't think thats the case (not sure userland would call connect with 0 sin_port that frequently).

**Comment by rscampos** [comment]
TL;DR
From my tests this is not a problem with Tracee, it's a normal behavior when application/lib uses `sin_port=0`. I was able to reproduce a similar behavior when `getaddrinfo()` is used to resolve a hostname.

I've been doing some tests and was able to reproduce the `sin_port=0`. This doesn't happen with frequency, but after I let Tracee run in my environment for some hours / day, I was able to see this behavior, and it happens over time. In my case, I was able to see this event on the `snapd` daemon process. In order to do a double check, also attached strace to capture the syscall `connect` of snapd.

Used the following Tracee command:
```
sudo ./dist/tracee -e security_socket_connect -o json
````
Tracee output:
```
00:58:42:316073  0      snapd            771016  771111  0                security_socket_connect   sockfd: 3, type: SOCK_DGRAM, remote_addr: map[sa_family:AF_INET sin_addr:185.125.188.58 sin_port:0]
00:58:42:316673  0      snapd            771016  771111  0                security_socket_connect   sockfd: 3, type: SOCK_DGRAM, remote_addr: map[sa_family:AF_INET sin_addr:185.125.188.55 sin_port:0]
06:13:42:337232  0      snapd            771016  771057  0                security_socket_connect   sockfd: 3, type: SOCK_DGRAM, remote_addr: map[sa_family:AF_INET sin_addr:185.125.188.55 sin_port:0]
06:13:42:338113  0      snapd            771016  771057  0                security_socket_connect   sockfd: 3, type: SOCK_DGRAM, remote_addr: map[sa_family:AF_INET sin_addr:185.125.188.58 sin_port:0]
06:13:42:338937  0      snapd            771016  771057  0                security_socket_connect   sockfd: 3, type: SOCK_DGRAM, remote_addr: map[sa_family:AF_INET sin_addr:185.125.188.59 sin_port:0]
```
Strace output:
````
[pid 771111] connect(3, {sa_family=AF_INET, sin_port=htons(0), sin_addr=inet_addr("185.125.188.58")}, 16) = 0
[pid 771111] connect(3, {sa_family=AF_INET, sin_port=htons(0), sin_addr=inet_addr("185.125.188.55")}, 16) = 0
[pid 771057] connect(3, {sa_family=AF_INET, sin_port=htons(0), sin_addr=inet_addr("185.125.188.55")}, 16) = 0
[pid 771057] connect(3, {sa_family=AF_INET, sin_port=htons(0), sin_addr=inet_addr("185.125.188.58")}, 16) = 0
[pid 771057] connect(3, {sa_family=AF_INET, sin_port=htons(0), sin_addr=inet_addr("185.125.188.59")}, 16) = 0
````
After doing some PoCs was able directly reproduce it using `wget www.youtube.com`:

```
18:24:37:813474  0      wget             1511059 1511059 0                security_socket_connect   sockfd: 3, type: SOCK_DGRAM, remote_addr: map[sa_family:AF_INET sin_addr:172.217.29.174 sin_port:0]
18:24:37:813536  0      wget             1511059 1511059 0                security_socket_connect   sockfd: 3, type: SOCK_DGRAM, remote_addr: map[sa_family:AF_INET sin_addr:172.217.29.142 sin_port:0]
18:24:37:813571  0      wget             1511059 1511059 0                security_socket_connect   sockfd: 3, type: SOCK_DGRAM, remote_addr: map[sa_family:AF_INET sin_addr:172.217.30.14 sin_port:0]
18:24:37:813603  0      wget             1511059 1511059 0                security_socket_connect   sockfd: 3, type: SOCK_DGRAM, remote_addr: map[sa_family:AF_INET sin_addr:172.217.30.46 sin_port:0]
18:24:37:813634  0      wget             1511059 1511059 0                security_socket_connect   sockfd: 3, type: SOCK_DGRAM, remote_addr: map[sa_family:AF_INET sin_addr:172.217.30.78 sin_port:0]
18:24:37:813666  0      wget             1511059 1511059 0                security_socket_connect   sockfd: 3, type: SOCK_DGRAM, remote_addr: map[sa_family:AF_INET sin_addr:142.250.218.206 sin_port:0]
18:24:37:813698  0      wget             1511059 1511059 0                security_socket_connect   sockfd: 3, type: SOCK_DGRAM, remote_addr: map[sa_family:AF_INET sin_addr:142.250.219.14 sin_port:0]
18:24:37:813730  0      wget             1511059 1511059 0                security_socket_connect   sockfd: 3, type: SOCK_DGRAM, remote_addr: map[sa_family:AF_INET sin_addr:172.217.29.238 sin_port:0]
```

Did some PoCs and after using the function `getaddrinfo(hostname,service,hints,res)` to resolve a hostname, was able to see `sin_port=0`. When used with `service=NULL`, after reading the IPs from the DNS response, `getaddrinfo` will call  `connect(sin_port=0)` for each IP returned. `connect()` on UDP can be used to check if destination could be reached by the route table, regardless of the destination port (`sin_port=0`).

So I was able to see this behavior using `wget` and call `getaddrinfo` directly in a C code. Also on `snapd` together with strace. I don't know if there is a specific corner case, but it seems that `sin_port=0` on Tracee is normal behavior.

If someone has another idea to test, pls let me know... I'd be happy to test it :)

**Comment by geyslan** [comment]
A zero `sin_port` communicates an "ephemeral port" to the kernel. It can be used in `getaddrinfo()` as you mentioned @rscampos and IIRC it makes the server `bind()` to any available port. You can check the actual port (assigned by the kernel) by calling `getsockname()`.

**Comment by yanivagman** [comment]
Thanks @rscampos for checking this.
This seems to be a legitimate usecase, so let's close it.
Please reopen if you think there is a real issue here.
